### ISSUE TITLE
Invalid netflow packet received

### ISSUE THREAD

**Issue by ruben-herold** [logstash-plugins/logstash-codec-netflow#128]
- Version:

  logstash: 6.2.3 
  logstash-codec-netflow: 3.11.2

- Operating System:
 Centos 7

- Config File (if you have sensitive info, please remove it):
 
   using supplied files from https://github.com/robcowart/elastiflow

- Steps to Reproduce:
  
   Log messages:

[2018-04-03T19:48:59,295][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:48:59,334][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:01,596][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:01,677][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:09,301][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:09,339][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:11,599][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:11,685][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:19,297][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:19,334][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:21,603][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-03T19:49:21,683][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)

**Comment by jorritfolmer** [comment]
Looks like you're receiving invalid netflow packets. Flowset_id can be 0,1,2 or 256->higher.
Perhaps some monitoring software or vulnerability scanner that does a scan?

**Comment by ruben-herold** [comment]
Logstash is direct connected to two juniper MX240 which sending the flows via IPv6 to the logstash.

**Comment by jorritfolmer** [comment]
Can you post a pcap of the netflow traffic?

**Comment by ruben-herold** [comment]
Yes I will attach a small one. 
But if I analyse it with wireshark I don't find anything with id 4. But during the capture time I received the log messages about id 4. 

[2018-04-04T10:20:19,779][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:19,875][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:19,877][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:19,921][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:24,896][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:25,005][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:25,009][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:25,073][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:29,779][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:29,872][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:29,875][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:29,919][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:32,202][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:32,339][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:32,343][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has be
en received. This message will usually go away after 1 minute.
[2018-04-04T10:20:32,361][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:34,533][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (End of file reached)
[2018-04-04T10:20:34,579][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (End of file reached)
[2018-04-04T10:20:39,781][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:39,898][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:39,900][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:39,931][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:42,206][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:42,307][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:42,320][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:42,335][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:49,262][INFO ][logstash.filters.translate] refreshing dictionary file
[2018-04-04T10:20:49,780][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:49,888][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:49,893][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 513 from observation domain id 590080, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:49,924][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)
[2018-04-04T10:20:52,588][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524544, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-04-04T10:20:52,756][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '4' not as expected for obj.records[1].flowset_id)

[ipfix.pcap.gz](https://github.com/logstash-plugins/logstash-codec-netflow/files/1874802/ipfix.pcap.gz)

**Comment by tiotzis73** [comment]
I have a similar problem with a value of 0.

[2018-04-04T13:26:31,097][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-04-04T13:26:46,563][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-04-04T13:26:59,397][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-04-04T13:26:59,420][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-04-04T13:27:25,176][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)

**Comment by ruben-herold** [comment]
@jorritfolmer  was the pcap ok or do you need some more?

**Comment by jorritfolmer** [comment]
Yes the pcap is fine. Unfortunately 'm unable to replay to Logstash on IP6.
Rewriting and replaying works because tcpdump shows them on the wire, on the Logstash host:

```
[root@host02 conf]# tcpdump -i eno16777736 -n port 2055 
13:48:15.842906 IP6 fd3d:5c2e:ed42:da88::1.50103 > fd3d:5c2e:ed42:da88::2.2055: UDP, length 72
13:48:17.692218 IP6 fd3d:5c2e:ed42:da88::1.50103 > fd3d:5c2e:ed42:da88::2.2055: UDP, length 80
```

Port is listening on udp6
```
[root@host02 conf]# netstat -plunt |grep 2055
udp6       0      0 :::2055                 :::*                                2324/java      
```

Addressing ok:
```
[root@host02 conf]# ip a sh |grep inet6
    inet6 ::1/128 scope host 
    inet6 fd3d:5c2e:ed42:da88::2/64 scope global 
    inet6 fe80::20c:29ff:feb9:5c8f/64 scope link 
```

And working because I can netcat some AAAA's to it (from host01)
```
echo "AAAAAAA"| nc -u fd3d:5c2e:ed42:da88::2 2055
```

Result (on host02):
````
[2018-04-06T13:54:07,098][WARN ][logstash.codecs.netflow  ] Ignoring Netflow version v16705
````

**Comment by ruben-herold** [comment]
looks similar to https://github.com/elastic/logstash/issues/3435

**Comment by kvicknair** [comment]
Using juniper MX-480 sending IPFIX:

[2018-05-18T11:38:49,609][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524288, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-05-18T11:38:49,654][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-05-18T11:38:50,859][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-05-18T11:38:59,905][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524288, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-05-18T11:38:59,969][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-05-18T11:39:09,735][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524288, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-05-18T11:39:09,755][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-05-18T11:39:19,682][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524288, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-05-18T11:39:19,732][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)
[2018-05-18T11:39:29,297][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524288, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-05-18T11:39:29,327][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)

**Comment by somedatapacket** [comment]
I am seeing this occur with records from an MX240 also, running JunOS 15.1R6-S3.

[2018-05-18T15:24:53,946][WARN ][logstash.codecs.netflow  ] Can't (yet) decode flowset id 512 from observation domain id 524832, because no template to decode it with has been received. This message will usually go away after 1 minute.
[2018-05-18T15:24:54,024][WARN ][logstash.codecs.netflow  ] Invalid netflow packet received (value '0' not as expected for obj.records[1].flowset_id)

**Comment by jorritfolmer** [comment]
@somedatapacket @kvicknair Can you post pcaps?

**Comment by jorritfolmer** [comment]
This is fixed in 4.0.1.
Thanks for the pcap!
